### ISSUE TITLE
Strange huge diff atol value when optimizing a model

### ISSUE THREAD

**Issue by jplu** [huggingface/optimum#978]
### System Info

```shell
optimum: 1.7.3
python: 3.10.10
OS: Linux Ubuntu 22.10
ORT-GPU: 1.14.1
Pytorch: 2.0.0
transformers: 4.27.4
```

### Who can help?

@michaelbenayoun @fx

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Here the command line I use:
```
optimum-cli export onnx -m cross-encoder/mmarco-mMiniLMv2-L12-H384-v1 --optimize O3 test/mmarco-mMiniLMv2-L12-H384-v1
```
Logs I get:
```
Framework not specified. Using pt to export to ONNX.
Automatic task detection to sequence-classification.
Using framework PyTorch: 2.0.0
Overriding 1 configuration item(s)
        - use_cache -> False
================ Diagnostic Run torch.onnx.export version 2.0.0 ================
verbose: False, log level: Level.ERROR
======================= 0 NONE 0 NOTE 0 WARNING 0 ERROR ========================

/home/jplu/miniconda3/envs/cluster-inference/lib/python3.10/site-packages/optimum/onnxruntime/configuration.py:735: FutureWarning: disable_embed_layer_norm will be deprecated soon, use disable_embed_layer_norm_fusion instead, disable_embed_layer_norm_fusion is set to True.
  warnings.warn(
Optimizing model...
symbolic shape inference disabled or failed.
Configuration saved in test/mmarco-mMiniLMv2-L12-H384-v1/ort_config.json
Optimized model saved at: test/mmarco-mMiniLMv2-L12-H384-v1 (external data format: False; saved all tensor to one file: True)
Validating ONNX model test/mmarco-mMiniLMv2-L12-H384-v1/model.onnx...
        -[✓] ONNX model output names match reference model (logits)
        - Validating ONNX Model output "logits":
                -[✓] (2, 1) matches (2, 1)
                -[x] values not close enough, max diff: 4.666891098022461 (atol: 0.0001)
The ONNX export succeeded with the warning: The maximum absolute difference between the output of the reference model and the ONNX exported model is not within the set tolerance 0.0001:
- logits: max diff = 4.666891098022461.
 The exported model was saved at: test/mmarco-mMiniLMv2-L12-H384-v1
```

I get an atol for 4.6 which is really huge. Any idea why? Same diff warning happens with `02` , `O4` and with `--device cuda  --fp16`.

### Expected behavior

Having a reasonable atol diff between the original model and the exported ONNX one.

**Comment by michaelbenayoun** [comment]
I do not have any issue on my side until `O3`. Something in the optimization seems to mess up with the accuracy.

**Comment by jplu** [comment]
Humm this is pretty strange. I did run it 10 times in a row and I always got a diff > 4 for the 10 runs. Anything I can check on my side to see if a setting or something is wrongly set?

**Comment by michaelbenayoun** [comment]
I do not observe those big differences. Thta's weird. Can you reinstlal optimum?

**Comment by fxmarty** [comment]
It could be an accuracy issue in an ORT op as well. If you'd like to investigate, it can be worth to compile ORT from source with the `ORT_DEBUG_NODE_IO_DUMP_OUTPUT_DATA=1` environment variable.

**Comment by jplu** [comment]
@michaelbenayoun I did reinstall optimum to 1.8.3dev0 and still the same issue with also 5 other models. Now it is even worse, I get atol diff of around 7.

@fxmarty I don't have the proper tools installed to build ORT on the machine I'm currently doing these tests and I'm not admin so it can take several days to have ORT compiled, as IT won't take my demand as a big priority. Any other thing I can do to help debugging this?

**Comment by jplu** [comment]
My CUDA version is 11.8, this might be the reason? Here I see that the CUDA version for 1.14 is 11.6.

**Comment by jplu** [comment]
I have tested on three different machines with brand new fresh python environment via conda, they all have a different configuration with different kind of GPU, and I still have the same issue on these three machines.

**Comment by fxmarty** [comment]
@jplu I would advise to open a reproducible issue on ORT repo with minimal dependency on optimum.

**Comment by jplu** [comment]
@fxmarty Ok thanks, I will try to figure this out, how to properly reproduce without optimum. 

Also, any chance you can share what is your env on which you are running this without issue, please? Such as ORT version (compiled or not?), optimum, Python and CUDA version? Maybe I can try the same.

**Comment by jplu** [comment]
@fxmarty I think there is an issue in the way optimum compute the diff. Here the small experiment I have done:
1. Create an ONNX export without optimization:
```
optimum-cli export onnx -m cross-encoder/mmarco-mMiniLMv2-L12-H384-v1 no_optim/
```
Here validation is ok:
```
Validating ONNX model test/mmarco-mMiniLMv2-L12-H384-v1/model.onnx...
        -[✓] ONNX model output names match reference model (logits)
        - Validating ONNX Model output "logits":
                -[✓] (2, 1) matches (2, 1)
                -[✓] all values close (atol: 0.0001)
```
2. Create an ONNX export with the O2 optimization:
```
optimum-cli export onnx -m cross-encoder/mmarco-mMiniLMv2-L12-H384-v1 --optimize O2 o2_optim/
```
Here the validation fails:
```
Validating ONNX model test/mmarco-mMiniLMv2-L12-H384-v1_o2/model.onnx...
        -[✓] ONNX model output names match reference model (logits)
        - Validating ONNX Model output "logits":
                -[✓] (2, 1) matches (2, 1)
                -[x] values not close enough, max diff: 6.75302267074585 (atol: 0.0001)
```
3. doing the diff manually:
```python
import onnxruntime as ort
import numpy as np
from transformers import AutoModelForSequenceClassification, AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("cross-encoder/mmarco-mMiniLMv2-L12-H384-v1")
model_pt = AutoModelForSequenceClassification.from_pretrained("cross-encoder/mmarco-mMiniLMv2-L12-H384-v1")
ort_sess_no_optim = ort.InferenceSession('no_optim/model.onnx', providers=['CPUExecutionProvider'])
ort_sess_o2 = ort.InferenceSession('o2_optim/model.onnx', providers=['CPUExecutionProvider'])
inputs_ort = tokenizer("bla", return_tensors='np')
inputs_pt = tokenizer("bla", return_tensors='pt')
outputs_no_optim = ort_sess_no_optim.run(None, dict(inputs_ort))
outputs_o2 = ort_sess_o2.run(None, dict(inputs_ort))
outputs_pt = model_pt(**inputs_pt).logits.detach().numpy()
print(f"Are ORT no optim and PT close enough? : {np.allclose(outputs_no_optim, outputs_pt, atol=1e-4)}")
print(f"Are ORT O2 and PT close enough? : {np.allclose(outputs_o2, outputs_pt, atol=1e-4)}")
print(f"Are ORT no optim and ORT O2 close enough? : {np.allclose(outputs_no_optim, outputs_o2, atol=1e-4)}")
```

And here the output I have:
```
Are ORT no optim and PT close enough? : True
Are ORT O2 and PT close enough? : True
Are ORT no optim and ORT O2 close enough? : True
```

They are all close, while when I do the convert with the O2 optim with the command line I get something very far to close.
